File: packages/ce-core/src/allocation.ts
/**
 * Kind-Aware Budget Allocation
 *
 * Replaces flat greedy packing with category-level budget constraints.
 * Instead of treating all items as a flat list, groups them by `kind`
 * and allocates budget per category with min/max/target constraints.
 *
 * Research: "Token-Budget-Aware LLM Reasoning" (ACL 2025) showed up to
 * 47% cost reduction with dynamic budget approaches while maintaining accuracy.
 */

import type { Budget, ContextItem, ContextPack, PackOptions } from "./types.js";
import { pack, packAsync } from "./pack.js";
import { KindAllocationSchema, validateWithSchema } from "./schemas.js";
import type { MaybeAsync } from "./maybe-async.js";
import { chain } from "./maybe-async.js";

/**
 * Budget constraint for a single item kind.
 */
export interface KindAllocation {
  /** The kind to match (e.g., "system", "memory", "retrieval") */
  kind: string;
  /** Minimum tokens to allocate (floor) */
  minTokens?: number;
  /** Maximum tokens to allocate (ceiling) */
  maxTokens?: number;
  /** Target fraction of total budget (0-1) */
  targetRatio?: number;
  /** Priority for surplus redistribution (higher = gets surplus first) */
  priority?: number;
}

/**
 * Per-kind allocation result.
 */
export interface KindResult {
  kind: string;
  budgetAllocated: number;
  budgetUsed: number;
  itemCount: number;
  surplus: number;
}

/**
 * Result of kind-aware packing, extending ContextPack.
 */
export interface AllocatedPack extends ContextPack {
  /** Per-kind allocation breakdown */
  allocations: Record<string, KindResult>;
  /** How well actual allocation matches target (0-1) */
  allocationEfficiency: number;
}

/** Function that packs items into a budget, returning sync or async. */
type PackFn = (
  items: ContextItem[],
  budget: Budget,
  options: PackOptions
) => MaybeAsync<ContextPack>;

/**
 * Shared implementation for both sync and async packWithAllocation.
 *
 * Uses MaybeAsync + chain() so the sync path never creates Promises,
 * while the async path chains naturally through .then().
 */
function packWithAllocationImpl(
  items: ContextItem[],
  budget: Budget,
  allocations: KindAllocation[],
  options: PackOptions,
  packFn: PackFn
): MaybeAsync<AllocatedPack> {
  for (let i = 0; i < allocations.length; i++) {
    validateWithSchema(
      KindAllocationSchema,
      allocations[i],
      `allocations[${i}]`
    );
  }

  const effectiveBudget = budget.maxTokens - (budget.reserveTokens ?? 0);
  const priorityByKind = new Map(
    allocations.map(alloc => [alloc.kind, alloc.priority ?? 0])
  );

  // Build a Set of allocated kinds for O(1) lookup (M1 fix)
  const allocatedKinds = new Set(allocations.map(a => a.kind));

  // Group items by kind
  const kindGroups = new Map<string, ContextItem[]>();
  const uncategorized: ContextItem[] = [];

  for (const item of items) {
    const kind = item.kind ?? "_uncategorized";
    if (allocatedKinds.has(kind)) {
      const group = kindGroups.get(kind) ?? [];
      group.push(item);
      kindGroups.set(kind, group);
    } else {
      uncategorized.push(item);
    }
  }

  // Phase 1: Compute initial allocation per kind
  const kindBudgets = new Map<string, number>();
  let allocatedTotal = 0;

  for (const alloc of allocations) {
    let tokens = 0;
    if (alloc.targetRatio !== undefined) {
      tokens = Math.floor(effectiveBudget * alloc.targetRatio);
    }
    if (alloc.minTokens !== undefined) {
      tokens = Math.max(tokens, alloc.minTokens);
    }
    if (alloc.maxTokens !== undefined) {
      tokens = Math.min(tokens, alloc.maxTokens);
    }
    kindBudgets.set(alloc.kind, tokens);
    allocatedTotal += tokens;
  }

  // Normalize if over-allocated. Prefer preserving higher-priority kinds.
  if (allocatedTotal > effectiveBudget) {
    let overflow = allocatedTotal - effectiveBudget;
    const adjustable = [...allocations].sort(
      (a, b) => (a.priority ?? 0) - (b.priority ?? 0)
    );

    for (const alloc of adjustable) {
      if (overflow <= 0) break;
      const current = kindBudgets.get(alloc.kind) ?? 0;
      const floor = alloc.minTokens ?? 0;
      const reducible = Math.max(0, current - floor);
      if (reducible <= 0) continue;
      const reduction = Math.min(reducible, overflow);
      kindBudgets.set(alloc.kind, current - reduction);
      overflow -= reduction;
    }

    for (const alloc of adjustable) {
      if (overflow <= 0) break;
      const current = kindBudgets.get(alloc.kind) ?? 0;
      if (current <= 0) continue;
      const reduction = Math.min(current, overflow);
      kindBudgets.set(alloc.kind, current - reduction);
      overflow -= reduction;
    }
  }

  // Phase 2: Pack within each kind's allocation
  const kindResults = new Map<
    string,
    { selected: ContextItem[]; dropped: ContextItem[]; used: number }
  >();
  let totalSurplus = 0;

  // Process each allocation sequentially, chaining MaybeAsync results
  let phase2: MaybeAsync<void> = undefined as unknown as void;

  for (const alloc of allocations) {
    phase2 = chain(phase2, () => {
      const kindItems = kindGroups.get(alloc.kind) ?? [];
      const kindBudget = kindBudgets.get(alloc.kind) ?? 0;

      if (kindItems.length === 0 || kindBudget <= 0) {
        kindResults.set(alloc.kind, {
          selected: [],
          dropped: kindItems,
          used: 0,
        });
        totalSurplus += kindBudget;
        return;
      }

      return chain(
        packFn(kindItems, { maxTokens: kindBudget }, options),
        result => {
          kindResults.set(alloc.kind, {
            selected: result.selected,
            dropped: result.dropped,
            used: result.totalTokens,
          });

          const surplus = kindBudget - result.totalTokens;
          if (surplus > 0) totalSurplus += surplus;
        }
      );
    });
  }

  // Phase 3: Redistribute surplus to kinds that need more space
  return chain(phase2, () => {
    let redistributeChain: MaybeAsync<void> = undefined as unknown as void;

    if (totalSurplus > 0) {
      const sortedByPriority = [...allocations].sort(
        (a, b) => (b.priority ?? 0) - (a.priority ?? 0)
      );

      for (const alloc of sortedByPriority) {
        redistributeChain = chain(redistributeChain, () => {
          if (totalSurplus <= 0) return;

          const result = kindResults.get(alloc.kind);
          if (!result || result.dropped.length === 0) return;

          const maxExtra =
            alloc.maxTokens !== undefined
              ? alloc.maxTokens - result.used
              : totalSurplus;

          if (maxExtra <= 0) return;

          const extraBudget = Math.min(totalSurplus, maxExtra);
          return chain(
            packFn(result.dropped, { maxTokens: extraBudget }, options),
            extraPack => {
              result.selected.push(...extraPack.selected);
              result.dropped = extraPack.dropped;
              result.used += extraPack.totalTokens;
              kindBudgets.set(
                alloc.kind,
                (kindBudgets.get(alloc.kind) ?? result.used) +
                  extraPack.totalTokens
              );
              totalSurplus -= extraPack.totalTokens;
            }
          );
        });
      }
    }

    // Phase 4: Pack uncategorized items into remaining budget
    return chain(redistributeChain, () => {
      const usedSoFar = Array.from(kindResults.values()).reduce(
        (sum, r) => sum + r.used,
        0
      );
      const remainingBudget = effectiveBudget - usedSoFar;

      const uncatPack: MaybeAsync<{
        selected: ContextItem[];
        dropped: ContextItem[];
      }> =
        uncategorized.length > 0 && remainingBudget > 0
          ? chain(
              packFn(uncategorized, { maxTokens: remainingBudget }, options),
              result => ({
                selected: result.selected,
                dropped: result.dropped,
              })
            )
          : { selected: [], dropped: uncategorized };

      return chain(uncatPack, uncategorizedResult => {
        // Compose final result
        const allSelected: ContextItem[] = [];
        const allDropped: ContextItem[] = [];
        const allocResult: Record<string, KindResult> = {};

        for (const alloc of allocations) {
          const result = kindResults.get(alloc.kind);
          if (!result) continue;
          allSelected.push(...result.selected);
          allDropped.push(...result.dropped);
          allocResult[alloc.kind] = {
            kind: alloc.kind,
            budgetAllocated: kindBudgets.get(alloc.kind) ?? 0,
            budgetUsed: result.used,
            itemCount: result.selected.length,
            surplus: Math.max(
              0,
              (kindBudgets.get(alloc.kind) ?? 0) - result.used
            ),
          };
        }

        allSelected.sort((a, b) => {
          const kindDelta =
            (priorityByKind.get(b.kind ?? "_uncategorized") ?? 0) -
            (priorityByKind.get(a.kind ?? "_uncategorized") ?? 0);
          if (kindDelta !== 0) return kindDelta;
          return (b.priority ?? 0) - (a.priority ?? 0);
        });

        allSelected.push(...uncategorizedResult.selected);
        allDropped.push(...uncategorizedResult.dropped);

        if (
          uncategorizedResult.selected.length > 0 ||
          uncategorizedResult.dropped.length > 0
        ) {
          const uncatTokens = uncategorizedResult.selected.reduce(
            (sum, i) => sum + (i.tokens ?? 0),
            0
          );
          allocResult["_uncategorized"] = {
            kind: "_uncategorized",
            budgetAllocated: remainingBudget,
            budgetUsed: uncatTokens,
            itemCount: uncategorizedResult.selected.length,
            surplus: Math.max(0, remainingBudget - uncatTokens),
          };
        }

        const totalTokens = allSelected.reduce(
          (sum, i) => sum + (i.tokens ?? 0),
          0
        );

        // Compute allocation efficiency: how close actual ratios match targets
        let efficiencySum = 0;
        let efficiencyCount = 0;
        for (const alloc of allocations) {
          if (alloc.targetRatio !== undefined && totalTokens > 0) {
            const actualRatio =
              (allocResult[alloc.kind]?.budgetUsed ?? 0) / totalTokens;
            const diff = Math.abs(actualRatio - alloc.targetRatio);
            efficiencySum += 1 - Math.min(diff / alloc.targetRatio, 1);
            efficiencyCount++;
          }
        }

        return {
          budget,
          selected: allSelected,
          dropped: allDropped,
          totalTokens,
          stats: {
            kindCount: allocations.length,
            remainingTokens: Math.max(0, effectiveBudget - totalTokens),
          },
          allocations: allocResult,
          allocationEfficiency:
            efficiencyCount > 0
              ? Math.round((efficiencySum / efficiencyCount) * 1000) / 1000
              : 1,
        };
      });
    });
  });
}

/**
 * Pack items with kind-aware budget allocation.
 *
 * Groups items by `kind`, allocates budget per category respecting
 * min/max/target constraints, then packs greedily within each allocation.
 * Surplus budget from underfilled categories is redistributed to
 * overfilled ones by priority.
 *
 * @param items - Context items to pack (should have `kind` set)
 * @param budget - Total token budget
 * @param allocations - Per-kind budget constraints
 * @param options - Standard pack options
 * @returns AllocatedPack with per-kind breakdown
 *
 * @example
 * ```ts
 * const result = packWithAllocation(
 *   items,
 *   { maxTokens: 8000 },
 *   [
 *     { kind: "system", minTokens: 500, maxTokens: 1500, targetRatio: 0.15, priority: 10 },
 *     { kind: "retrieval", targetRatio: 0.40, priority: 5 },
 *     { kind: "conversation", targetRatio: 0.35, priority: 7 },
 *     { kind: "memory", targetRatio: 0.10, priority: 3 },
 *   ],
 *   {}
 * );
 * console.log(result.allocations);
 * console.log(result.allocationEfficiency);
 * ```
 */
export function packWithAllocation(
  items: ContextItem[],
  budget: Budget,
  allocations: KindAllocation[],
  options: PackOptions = {}
): AllocatedPack {
  return packWithAllocationImpl(
    items,
    budget,
    allocations,
    options,
    pack
  ) as AllocatedPack;
}

/**
 * Async variant of packWithAllocation.
 *
 * Same logic as packWithAllocation but delegates to packAsync() internally,
 * supporting async operations like embedding-based redundancy elimination.
 *
 * @param items - Context items to pack (should have `kind` set)
 * @param budget - Total token budget
 * @param allocations - Per-kind budget constraints
 * @param options - Standard pack options (supports embeddingProvider, redundancyConfig)
 * @returns Promise<AllocatedPack> with per-kind breakdown
 */
export async function packWithAllocationAsync(
  items: ContextItem[],
  budget: Budget,
  allocations: KindAllocation[],
  options: PackOptions = {}
): Promise<AllocatedPack> {
  return packWithAllocationImpl(
    items,
    budget,
    allocations,
    options,
    packAsync
  ) as Promise<AllocatedPack>;
}

File: python/context_engineering/allocation.py
"""
Kind-Aware Budget Allocation

Replaces flat greedy packing with category-level budget constraints.
Instead of treating all items as a flat list, groups them by `kind`
and allocates budget per category with min/max/target constraints.

Research: "Token-Budget-Aware LLM Reasoning" (ACL 2025) showed up to
47% cost reduction with dynamic budget approaches while maintaining accuracy.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, List, Optional

from .core import Budget, ContextItem, pack, pack_async
from .errors import ValidationError


@dataclass
class KindAllocation:
    """Budget constraint for a single item kind."""

    kind: str
    min_tokens: Optional[int] = None
    max_tokens: Optional[int] = None
    target_ratio: Optional[float] = None
    priority: int = 0


@dataclass
class KindResult:
    """Per-kind allocation result."""

    kind: str
    budget_allocated: int
    budget_used: int
    item_count: int
    surplus: int


@dataclass
class AllocatedPack:
    """Result of kind-aware packing."""

    budget: Budget
    selected: List[ContextItem]
    dropped: List[ContextItem]
    total_tokens: int
    stats: Dict[str, Any]
    allocations: Dict[str, KindResult]
    allocation_efficiency: float


def pack_with_allocation(
    items: List[ContextItem],
    budget: Budget,
    allocations: List[KindAllocation],
    options: Optional[Dict[str, Any]] = None,
) -> AllocatedPack:
    """Pack items with kind-aware budget allocation.

    Groups items by `kind`, allocates budget per category respecting
    min/max/target constraints, then packs greedily within each allocation.
    Surplus budget from underfilled categories is redistributed to
    overfilled ones by priority.

    Args:
        items: Context items to pack (should have `kind` set)
        budget: Total token budget
        allocations: Per-kind budget constraints
        options: Standard pack options

    Returns:
        AllocatedPack with per-kind breakdown

    Example:
        result = pack_with_allocation(
            items,
            Budget(max_tokens=8000),
            [
                KindAllocation(kind="system", min_tokens=500, target_ratio=0.15, priority=10),
                KindAllocation(kind="retrieval", target_ratio=0.40, priority=5),
                KindAllocation(kind="conversation", target_ratio=0.35, priority=7),
            ],
        )
        print(result.allocations)
        print(result.allocation_efficiency)
    """
    details = []
    for i, alloc in enumerate(allocations):
        kind = alloc.kind if hasattr(alloc, "kind") else alloc.get("kind", "")
        if not kind:
            details.append(
                {
                    "path": f"allocations[{i}].kind",
                    "message": "kind must be a non-empty string",
                }
            )
        ratio = (
            alloc.target_ratio
            if hasattr(alloc, "target_ratio")
            else alloc.get("target_ratio", None)
        )
        if ratio is not None and not (0 <= ratio <= 1):
            details.append(
                {
                    "path": f"allocations[{i}].target_ratio",
                    "message": "target_ratio must be between 0 and 1",
                }
            )
    if details:
        raise ValidationError("Invalid allocation config", details=details)

    pack_kwargs = dict(options or {})
    effective_budget = budget.max_tokens - (budget.reserve_tokens or 0)

    # Group items by kind
    kind_groups: Dict[str, List[ContextItem]] = {}
    uncategorized: List[ContextItem] = []
    alloc_kinds = {a.kind for a in allocations}

    for item in items:
        kind = item.kind or "_uncategorized"
        if kind in alloc_kinds:
            kind_groups.setdefault(kind, []).append(item)
        else:
            uncategorized.append(item)

    # Phase 1: Compute initial allocation per kind
    kind_budgets: Dict[str, int] = {}
    allocated_total = 0

    for alloc in allocations:
        tokens = 0
        if alloc.target_ratio is not None:
            tokens = int(effective_budget * alloc.target_ratio)
        if alloc.min_tokens is not None:
            tokens = max(tokens, alloc.min_tokens)
        if alloc.max_tokens is not None:
            tokens = min(tokens, alloc.max_tokens)
        kind_budgets[alloc.kind] = tokens
        allocated_total += tokens

    # Scale if over-allocated
    if allocated_total > effective_budget:
        scale = effective_budget / allocated_total
        for alloc in allocations:
            scaled = int(kind_budgets[alloc.kind] * scale)
            if alloc.min_tokens is not None:
                scaled = max(scaled, alloc.min_tokens)
            kind_budgets[alloc.kind] = scaled

    # Phase 2: Pack within each kind's allocation
    kind_results: Dict[str, Dict[str, Any]] = {}
    total_surplus = 0

    for alloc in allocations:
        kind_items = kind_groups.get(alloc.kind, [])
        kind_budget = kind_budgets.get(alloc.kind, 0)

        if not kind_items or kind_budget <= 0:
            kind_results[alloc.kind] = {"selected": [], "dropped": list(kind_items), "used": 0}
            total_surplus += kind_budget
            continue

        result = pack(kind_items, Budget(max_tokens=kind_budget), **pack_kwargs)
        kind_results[alloc.kind] = {
            "selected": list(result.selected),
            "dropped": list(result.dropped),
            "used": result.total_tokens,
        }

        surplus = kind_budget - result.total_tokens
        if surplus > 0:
            total_surplus += surplus

    # Phase 3: Redistribute surplus to kinds that need more space
    if total_surplus > 0:
        sorted_allocs = sorted(allocations, key=lambda a: a.priority, reverse=True)

        for alloc in sorted_allocs:
            if total_surplus <= 0:
                break

            result = kind_results.get(alloc.kind)
            if not result or not result["dropped"]:
                continue

            max_extra = (
                alloc.max_tokens - result["used"] if alloc.max_tokens is not None else total_surplus
            )
            if max_extra <= 0:
                continue

            extra_budget = min(total_surplus, max_extra)
            extra_pack = pack(result["dropped"], Budget(max_tokens=extra_budget), **pack_kwargs)

            result["selected"].extend(extra_pack.selected)
            result["dropped"] = list(extra_pack.dropped)
            result["used"] += extra_pack.total_tokens
            total_surplus -= extra_pack.total_tokens

    # Phase 4: Pack uncategorized items into remaining budget
    used_so_far = sum(r["used"] for r in kind_results.values())
    remaining_budget = effective_budget - used_so_far

    if uncategorized and remaining_budget > 0:
        uncat_result = pack(uncategorized, Budget(max_tokens=remaining_budget), **pack_kwargs)
        uncat_selected = list(uncat_result.selected)
        uncat_dropped = list(uncat_result.dropped)
    else:
        uncat_selected = []
        uncat_dropped = list(uncategorized)

    # Compose final result
    all_selected: List[ContextItem] = []
    all_dropped: List[ContextItem] = []
    alloc_result: Dict[str, KindResult] = {}

    for alloc in allocations:
        result = kind_results[alloc.kind]
        all_selected.extend(result["selected"])
        all_dropped.extend(result["dropped"])
        alloc_result[alloc.kind] = KindResult(
            kind=alloc.kind,
            budget_allocated=kind_budgets.get(alloc.kind, 0),
            budget_used=result["used"],
            item_count=len(result["selected"]),
            surplus=max(0, kind_budgets.get(alloc.kind, 0) - result["used"]),
        )

    all_selected.extend(uncat_selected)
    all_dropped.extend(uncat_dropped)

    if uncat_selected or uncat_dropped:
        uncat_tokens = sum(i.tokens or 0 for i in uncat_selected)
        alloc_result["_uncategorized"] = KindResult(
            kind="_uncategorized",
            budget_allocated=remaining_budget,
            budget_used=uncat_tokens,
            item_count=len(uncat_selected),
            surplus=max(0, remaining_budget - uncat_tokens),
        )

    total_tokens = sum(i.tokens or 0 for i in all_selected)

    # Compute allocation efficiency
    efficiency_sum = 0.0
    efficiency_count = 0
    for alloc in allocations:
        if alloc.target_ratio is not None and total_tokens > 0:
            actual_ratio = (alloc_result[alloc.kind].budget_used) / total_tokens
            diff = abs(actual_ratio - alloc.target_ratio)
            if alloc.target_ratio > 0:
                efficiency_sum += 1 - min(diff / alloc.target_ratio, 1)
            else:
                efficiency_sum += 1.0  # perfect efficiency if no target
            efficiency_count += 1

    return AllocatedPack(
        budget=budget,
        selected=all_selected,
        dropped=all_dropped,
        total_tokens=total_tokens,
        stats={
            "kindCount": len(allocations),
            "remainingTokens": max(0, effective_budget - total_tokens),
        },
        allocations=alloc_result,
        allocation_efficiency=round(efficiency_sum / efficiency_count, 3)
        if efficiency_count > 0
        else 1.0,
    )


def _extract_async_kwargs(options: Optional[Dict[str, Any]]) -> Dict[str, Any]:
    """Extract keyword arguments compatible with pack_async from options dict."""
    if not options:
        return {}
    kwargs: Dict[str, Any] = {}
    if "weights" in options:
        kwargs["weights"] = options["weights"]
    if "redundancy_config" in options:
        kwargs["redundancy_config"] = options["redundancy_config"]
    if "provider" in options:
        kwargs["provider"] = options["provider"]
    if "allow_compression" in options:
        kwargs["allow_compression"] = options["allow_compression"]
    return kwargs


async def pack_with_allocation_async(
    items: List[ContextItem],
    budget: Budget,
    allocations: List[KindAllocation],
    options: Optional[Dict[str, Any]] = None,
) -> AllocatedPack:
    """Async version of pack_with_allocation that uses pack_async internally.

    Supports asynchronous redundancy elimination via pack_async.
    See pack_with_allocation for full documentation.

    Args:
        items: Context items to pack (should have `kind` set)
        budget: Total token budget
        allocations: Per-kind budget constraints
        options: Standard pack options

    Returns:
        AllocatedPack with per-kind breakdown
    """
    details = []
    for i, alloc in enumerate(allocations):
        kind = alloc.kind if hasattr(alloc, "kind") else alloc.get("kind", "")
        if not kind:
            details.append(
                {
                    "path": f"allocations[{i}].kind",
                    "message": "kind must be a non-empty string",
                }
            )
        ratio = (
            alloc.target_ratio
            if hasattr(alloc, "target_ratio")
            else alloc.get("target_ratio", None)
        )
        if ratio is not None and not (0 <= ratio <= 1):
            details.append(
                {
                    "path": f"allocations[{i}].target_ratio",
                    "message": "target_ratio must be between 0 and 1",
                }
            )
    if details:
        raise ValidationError("Invalid allocation config", details=details)

    async_kwargs = _extract_async_kwargs(options)
    effective_budget = budget.max_tokens - (budget.reserve_tokens or 0)

    # Group items by kind
    kind_groups: Dict[str, List[ContextItem]] = {}
    uncategorized: List[ContextItem] = []
    alloc_kinds = {a.kind for a in allocations}

    for item in items:
        kind = item.kind or "_uncategorized"
        if kind in alloc_kinds:
            kind_groups.setdefault(kind, []).append(item)
        else:
            uncategorized.append(item)

    # Phase 1: Compute initial allocation per kind
    kind_budgets: Dict[str, int] = {}
    allocated_total = 0

    for alloc in allocations:
        tokens = 0
        if alloc.target_ratio is not None:
            tokens = int(effective_budget * alloc.target_ratio)
        if alloc.min_tokens is not None:
            tokens = max(tokens, alloc.min_tokens)
        if alloc.max_tokens is not None:
            tokens = min(tokens, alloc.max_tokens)
        kind_budgets[alloc.kind] = tokens
        allocated_total += tokens

    # Scale if over-allocated
    if allocated_total > effective_budget:
        scale = effective_budget / allocated_total
        for alloc in allocations:
            scaled = int(kind_budgets[alloc.kind] * scale)
            if alloc.min_tokens is not None:
                scaled = max(scaled, alloc.min_tokens)
            kind_budgets[alloc.kind] = scaled

    # Phase 2: Pack within each kind's allocation
    kind_results: Dict[str, Dict[str, Any]] = {}
    total_surplus = 0

    for alloc in allocations:
        kind_items = kind_groups.get(alloc.kind, [])
        kind_budget = kind_budgets.get(alloc.kind, 0)

        if not kind_items or kind_budget <= 0:
            kind_results[alloc.kind] = {"selected": [], "dropped": list(kind_items), "used": 0}
            total_surplus += kind_budget
            continue

        result = await pack_async(kind_items, Budget(max_tokens=kind_budget), **async_kwargs)
        kind_results[alloc.kind] = {
            "selected": list(result.selected),
            "dropped": list(result.dropped),
            "used": result.total_tokens,
        }

        surplus = kind_budget - result.total_tokens
        if surplus > 0:
            total_surplus += surplus

    # Phase 3: Redistribute surplus to kinds that need more space
    if total_surplus > 0:
        sorted_allocs = sorted(allocations, key=lambda a: a.priority, reverse=True)

        for alloc in sorted_allocs:
            if total_surplus <= 0:
                break

            result = kind_results.get(alloc.kind)
            if not result or not result["dropped"]:
                continue

            max_extra = (
                alloc.max_tokens - result["used"] if alloc.max_tokens is not None else total_surplus
            )
            if max_extra <= 0:
                continue

            extra_budget = min(total_surplus, max_extra)
            extra_pack = await pack_async(
                result["dropped"], Budget(max_tokens=extra_budget), **async_kwargs
            )

            result["selected"].extend(extra_pack.selected)
            result["dropped"] = list(extra_pack.dropped)
            result["used"] += extra_pack.total_tokens
            total_surplus -= extra_pack.total_tokens

    # Phase 4: Pack uncategorized items into remaining budget
    used_so_far = sum(r["used"] for r in kind_results.values())
    remaining_budget = effective_budget - used_so_far

    if uncategorized and remaining_budget > 0:
        uncat_result = await pack_async(
            uncategorized, Budget(max_tokens=remaining_budget), **async_kwargs
        )
        uncat_selected = list(uncat_result.selected)
        uncat_dropped = list(uncat_result.dropped)
    else:
        uncat_selected = []
        uncat_dropped = list(uncategorized)

    # Compose final result
    all_selected: List[ContextItem] = []
    all_dropped: List[ContextItem] = []
    alloc_result: Dict[str, KindResult] = {}

    for alloc in allocations:
        result = kind_results[alloc.kind]
        all_selected.extend(result["selected"])
        all_dropped.extend(result["dropped"])
        alloc_result[alloc.kind] = KindResult(
            kind=alloc.kind,
            budget_allocated=kind_budgets.get(alloc.kind, 0),
            budget_used=result["used"],
            item_count=len(result["selected"]),
            surplus=max(0, kind_budgets.get(alloc.kind, 0) - result["used"]),
        )

    all_selected.extend(uncat_selected)
    all_dropped.extend(uncat_dropped)

    if uncat_selected or uncat_dropped:
        uncat_tokens = sum(i.tokens or 0 for i in uncat_selected)
        alloc_result["_uncategorized"] = KindResult(
            kind="_uncategorized",
            budget_allocated=remaining_budget,
            budget_used=uncat_tokens,
            item_count=len(uncat_selected),
            surplus=max(0, remaining_budget - uncat_tokens),
        )

    total_tokens = sum(i.tokens or 0 for i in all_selected)

    # Compute allocation efficiency
    efficiency_sum = 0.0
    efficiency_count = 0
    for alloc in allocations:
        if alloc.target_ratio is not None and total_tokens > 0:
            actual_ratio = (alloc_result[alloc.kind].budget_used) / total_tokens
            diff = abs(actual_ratio - alloc.target_ratio)
            if alloc.target_ratio > 0:
                efficiency_sum += 1 - min(diff / alloc.target_ratio, 1)
            else:
                efficiency_sum += 1.0
            efficiency_count += 1

    return AllocatedPack(
        budget=budget,
        selected=all_selected,
        dropped=all_dropped,
        total_tokens=total_tokens,
        stats={
            "kindCount": len(allocations),
            "remainingTokens": max(0, effective_budget - total_tokens),
        },
        allocations=alloc_result,
        allocation_efficiency=round(efficiency_sum / efficiency_count, 3)
        if efficiency_count > 0
        else 1.0,
    )
